src/Core/Resolvers/Factories/MutationEngineFactory.cs

```diff
@@ -0,0 +1,77 @@
+using Azure.DataApiBuilder.Core.Services.MetadataProviders;
+using Azure.DataApiBuilder.Service.Exceptions;
+using Microsoft.AspNetCore.Http;
+
+namespace Azure.DataApiBuilder.Core.Resolvers.Factories
+{
+    /// <summary>
+    /// MutationEngineFactory class.
+    /// Used to get the IMutationEngine based on database type.
+    /// </summary>
+    public class MutationEngineFactory : IMutationEngineFactory
+    {
+        private readonly IEnumerable<IMutationEngine> _mutationEngines;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="MutationEngineFactory"/> class.
+        /// </summary>
+        /// <param name="runtimeConfigProvider">runtimeConfigProvider.</param>
+        /// <param name="queryManagerFactory">queryManagerFactory</param>
+        /// <param name="metadataProviderFactory">metadataProviderFactory.</param>
+        /// <param name="cosmosClientProvider">cosmosClientProvider</param>
+        /// <param name="queryEngineFactory">queryEngineFactory.</param>
+        /// <param name="httpContextAccessor">httpContextAccessor.</param>
+        /// <param name="authorizationResolver">authorizationResolver.</param>
+        /// <param name="gQLFilterParser">GqlFilterParser.</param>
+        public MutationEngineFactory(RuntimeConfigProvider runtimeConfigProvider,
+            IAbstractQueryManagerFactory queryManagerFactory,
+            IMetadataProviderFactory metadataProviderFactory,
+            CosmosClientProvider cosmosClientProvider,
+            IQueryEngineFactory queryEngineFactory,
+            IHttpContextAccessor httpContextAccessor,
+            IAuthorizationResolver authorizationResolver,
+            GQLFilterParser gQLFilterParser)
+        {
+            _mutationEngines = new List<IMutationEngine>();
+
+            RuntimeConfig config = runtimeConfigProvider.GetConfig();
+
+            if (config.SqlEngineNeeded)
+            {
+                _mutationEngines = _mutationEngines.Append(
+                    new SqlMutationEngine(
+                        queryManagerFactory, metadataProviderFactory, queryEngineFactory, authorizationResolver, gQLFilterParser, httpContextAccessor, runtimeConfigProvider));
+            }
+
+            if (config.CosmosEngineNeeded)
+            {
+                _mutationEngines = _mutationEngines.Append(
+                    new CosmosMutationEngine(cosmosClientProvider, metadataProviderFactory, authorizationResolver));
+            }
+        }
+
+        /// <inheritdoc/>
+        public IMutationEngine GetMutationEngine(DatabaseType databaseType)
+        {
+            IMutationEngine mutationEngine = databaseType switch
+            {
+                DatabaseType.CosmosDB_NoSQL => _mutationEngines.First(engine => engine.GetType() == typeof(CosmosMutationEngine)),
+                DatabaseType.MySQL or DatabaseType.MSSQL or DatabaseType.PostgreSQL => _mutationEngines.First(engine => engine.GetType() == typeof(SqlMutationEngine)),
```

Comment:
if there is only one engine for cosmos and one engine for sql, why not use a dictionary to return the engine that you want? 
key: cosmos  value: cosmosEngine
key: mssql value: sqlEngine
key: pgsql value: (same engine instance as mssql key)

because `_mutationEngines` is an implementation detail of MutationEngineFactory, I don't see the benefit of the `Ienumerable` abstraction.

Using dictionary would also help you avoid the extra work (allocs/checks/enumerable traversal) that comes with `.First()`

Reply:
Yes only one engine instance. I can change this to dictionary. Since the number of db types were constant and only 4 did not feel there would be a huge perf hit of using the IEnumerable.

Reply:
fixed.

Reply:
> Yes only one engine instance. I can change this to dictionary. Since the number of db types were constant and only 4 did not feel there would be a huge perf hit of using the IEnumerable.

It depends on the frequency of calls to `GetMutationEngine()`. It seems like this would be called for every mutation request (GraphQL/REST) so the number of allocations would be proportional to the number of times the mutationEngine is requested. Maybe a micro optimization since we don't know the amount of load DAB may be subject to. But good to just address this now before we need to go hunting for places in the codebase to optimize.